### ISSUE TITLE
key_exists crash in ejabberd_bosh.erl

### ISSUE THREAD

**Issue by jploski** [processone/ejabberd#2277]
> What version of ejabberd are you using?

17.12

> What operating system (version) are you using?

Debian Linux (jessie-slim)

> How did you install ejabberd (source, package, distribution)?

Compiled from source, via Dockerfile adapted from the one in Git repository (the Git repository version did not work for build, required inclusion of jessie-backports for a newer OTC and a few other fixes).

> What did not work as expected? Are there error messages in the log? What
> was the unexpected behavior? What was the expected result?

Often (not always) when attempting to join a MUC chatroom (using Converse.js client) the client appears to "hang". On the server the following crash message is logged. Afterwards the client reports dropped connection. Eventually, by refreshing the page, the same room may load without crash. Perhaps relevant: the room is configured with XEP-0313  MAM (the client uses MAM v.0.6.1). I did not manage to reproduce this using a different client than Converse.js - so it may be also related to connection management, the client opening multiple connections with same identity or some such.

```
2018-02-13 15:17:57 =ERROR REPORT====
** State machine <0.753.0> terminating
** Last message in was {'$gen_sync_event',{<0.759.0>,#Ref<0.0.2.4805>},{body,<<>>,[{sid,<<"8be04a6a55dde74ca080b13eaa955294f56c24ff">>},{rid,4066956713}],[],115}}
** When State == active
**      Data  == {state,<<"drpower.hea.com">>,<<"8be04a6a55dde74ca080b13eaa955294f56c24ff">>,{{[],[]},0,unlimited},{{[],[]},0,unlimited},none,<0.754.0>,<<"1.0">>,undefined,#Ref<0.0.2.4794>,60,30,4066956720,<<>>,undefined,unlimited,{2,{4066956719,{body,<<>>,[],[{xmlstreamelement,{xmlel,<<"presence">>,[{<<"xml:lang">>,<<"en">>},{<<"to">>,<<"jpl@drpower.hea.com/converse.js-57708534">>},{<<"from">>,<<"visitor.1905@conference.drpower.hea.com/Jan Ploski">>},{<<"type">>,<<"unavailable">>},{<<"id">>,<<"f53e6959-f591-4f1c-9028-0592710cebdf:sendPresence">>}],[{xmlel,<<"x">>,[{<<"xmlns">>,<<"http://jabber.org/protocol/muc#user">>}],[{xmlel,<<"item">>,[{<<"jid">>,<<"jpl@drpower.hea.com/converse.js-57708534">>},{<<"role">>,<<"none">>},{<<"affiliation">>,<<"owner">>}],[]},{xmlel,<<"status">>,[{<<"code">>,<<"110">>}],[]}]},{xmlel,<<"status">>,[],[]}]}}],0},nil,{4066956720,{body,<<>>,[],[],0},nil,nil}}},{3,{4066956712,{{<0.744.0>,#Ref<0.0.2.4711>},{body,<<>>,[{sid,<<"8be04a6a55dde74ca080b13eaa955294f56c24ff">>},{rid,4066956712}],[],115}},nil,{4066956713,{{<0.743.0>,#Ref<0.0.2.4786>},{body,<<>>,[{sid,<<"8be04a6a55dde74ca080b13eaa955294f56c24ff">>},{rid,4066956713}],[],115}},nil,{4066956714,{{<0.758.0>,#Ref<0.0.2.4793>},{body,<<>>,[{sid,<<"8be04a6a55dde74ca080b13eaa955294f56c24ff">>},{rid,4066956714}],[{xmlstreamelement,{xmlel,<<"iq">>,[{<<"xmlns">>,<<"jabber:client">>},{<<"to">>,<<"conference.drpower.hea.com">>},{<<"from">>,<<"jpl@drpower.hea.com/converse.js-57708534">>},{<<"type">>,<<"get">>},{<<"id">>,<<"26525879-3599-45f1-8bb9-f20c5cd5a1ab:sendIQ">>}],[{xmlel,<<"query">>,[{<<"xmlns">>,<<"http://jabber.org/protocol/disco#items">>}],[]}]}},{xmlstreamelement,{xmlel,<<"presence">>,[{<<"xmlns">>,<<"jabber:client">>},{<<"type">>,<<"unavailable">>},{<<"from">>,<<"jpl@drpower.hea.com/converse.js-57708534">>},{<<"to">>,<<"rypl-dispatch@conference.drpower.hea.com/Jan Ploski">>},{<<"id">>,<<"8691bc0f-c4fd-436e-abba-a9399b477f92:sendPresence">>}],[{xmlel,<<"status">>,[],[]}]}},{xmlstreamelement,{xmlel,<<"iq">>,[{<<"xmlns">>,<<"jabber:client">>},{<<"from">>,<<"jpl@drpower.hea.com/converse.js-57708534">>},{<<"to">>,<<"visitor.1905@conference.drpower.hea.com">>},{<<"type">>,<<"get">>},{<<"id">>,<<"d9c39cc8-ab61-4346-b715-50822f413544:sendIQ">>}],[{xmlel,<<"query">>,[{<<"xmlns">>,<<"http://jabber.org/protocol/disco#info">>}],[]}]}}],816}},nil,nil}}}},{{[],[]},0,1000},{{0,0,0,0,0,65535,32512,1},33104},2}
** Reason for termination =
** {{key_exists,4066956713},[{gb_trees,insert_1,4,[{file,"gb_trees.erl"},{line,318}]},{gb_trees,insert_1,4,[{file,"gb_trees.erl"},{line,297}]},{gb_trees,insert,3,[{file,"gb_trees.erl"},{line,277}]},{ejabberd_bosh,active1,3,[{file,"src/ejabberd_bosh.erl"},{line,465}]},{p1_fsm,handle_msg,10,[{file,"src/p1_fsm.erl"},{line,582}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
2018-02-13 15:17:57 =CRASH REPORT====
  crasher:
    initial call: ejabberd_bosh:init/1
    pid: <0.753.0>
    registered_name: []
    exception exit: {{{key_exists,4066956713},[{gb_trees,insert_1,4,[{file,"gb_trees.erl"},{line,318}]},{gb_trees,insert_1,4,[{file,"gb_trees.erl"},{line,297}]},{gb_trees,insert,3,[{file,"gb_trees.erl"},{line,277}]},{ejabberd_bosh,active1,3,[{file,"src/ejabberd_bosh.erl"},{line,465}]},{p1_fsm,handle_msg,10,[{file,"src/p1_fsm.erl"},{line,582}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]},[{p1_fsm,terminate,8,[{file,"src/p1_fsm.erl"},{line,760}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
    ancestors: ['mod_bosh_drpower.hea.com',ejabberd_gen_mod_sup,ejabberd_sup,<0.62.0>]
    messages: []
    links: [<0.664.0>]
    dictionary: [{'$internal_queue_len',0}]
    trap_exit: false
    status: running
    heap_size: 6772
    stack_size: 27
    reductions: 4546
  neighbours:
2018-02-13 15:17:57 =SUPERVISOR REPORT====
     Supervisor: {local,'mod_bosh_drpower.hea.com'}
     Context:    child_terminated
     Reason:     {{key_exists,4066956713},[{gb_trees,insert_1,4,[{file,"gb_trees.erl"},{line,318}]},{gb_trees,insert_1,4,[{file,"gb_trees.erl"},{line,297}]},{gb_trees,insert,3,[{file,"gb_trees.erl"},{line,277}]},{ejabberd_bosh,active1,3,[{file,"src/ejabberd_bosh.erl"},{line,465}]},{p1_fsm,handle_msg,10,[{file,"src/p1_fsm.erl"},{line,582}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
     Offender:   [{pid,<0.753.0>},{id,undefined},{mfargs,{ejabberd_bosh,start_link,undefined}},{restart_type,temporary},{shutdown,1000},{child_type,worker}]
```

**Comment by zinid** [comment]
Yeah, we have seen this crash before but we cannot reproduce it, partially, because the algorithm defined in the BOSH spec is very obscure.

**Comment by jploski** [comment]
I can reproduce it fairly easily on my development server (albeit only with Converse.js client).

Maybe the [crash and debug log](https://dev.plosquare.com/ejabberd/crash-and-debug-log-issue2277.tar.gz) captured during the event will help understand it further? I suspect that 522903848 in the crash message refers to a request id and that the client may be sending duplicate request ids, triggering some race condition in ejabberd. I'm not familiar with Erlang, but I can insert additional checks/debug outputs as instructed.

**Comment by zinid** [comment]
Well, as far as I remember from previous investigation, yes, it's somehow related to incorrect IDs used by a client. And if I remember correctly this is fatal situation, so we probably need to close the connection anyway.
I will look into the log when I'll be in mood, thanks.

**Comment by prefiks** [comment]
This should be fixed by commit 548ef7b835fbf2ad965bb4bb3a6f4e4905717002
